File: src/solutions/2023/09/ts/main.ts
import { argv } from 'node:process';
import type { SolutionOptions } from 'src/options.ts';
import { run } from 'src/run.ts';

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function difference(ary: number[], sz: number): number[] {
   for (let i = 0; i < sz; i++) ary[i] = ary[i + 1] - ary[i];
   return ary;
}

function extrapolate(ary: number[], sz: number): number {
   sz--;
   const last = ary[sz];
   if (!sz) return last;
   return extrapolate(difference(ary, sz), sz) + last;
}

function parseInput(input: string): number[][] {
   return input.split('\n').map((str) => str.split(' ').map(Number));
}

function part1(input: string, _isTest: boolean): string {
   const parsed = parseInput(input);
   let res = 0;
   for (const history of parsed) res += extrapolate(history, history.length);
   return res.toString();
}

function part2(input: string, _isTest: boolean): string {
   const parsed = parseInput(input);
   let res = 0;
   for (const history of parsed) res += extrapolate(history.reverse(), history.length);
   return res.toString();
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}

File: src/solutions/2023/09/csharp/Main.cs

using Core;

namespace Year2023;

public static class Day09
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = false,
      HasIO = false,
   };

   static long[] Difference(long[] ary, uint sz)
   {
      for (int i = 0; i < sz; i++)
      {
         ary[i] = ary[i + 1] - ary[i];
      }
      return ary;
   }

   static long Extrapolate(long[] ary, uint sz)
   {
      sz--;
      var last = ary[sz];
      if (sz == 0)
         return last;
      return Extrapolate(Difference(ary, sz), sz) + last;
   }

   static string Part1(string input, bool isTest = false)
   {
      return input.Split('\n')
         .Select(x =>
            Extrapolate(x
               .Split(' ')
               .Select(s => long.Parse(s))
               .ToArray(), (uint)x.Split(' ').Length))
         .Sum()
         .ToString();
   }

   static string Part2(string input, bool isTest = false)
   {
      return input.Split('\n')
         .Select(x =>
            Extrapolate(
               x
               .Split(' ')
               .Select(s => long.Parse(s))
               .Reverse()
               .ToArray(), (uint)x.Split(' ').Length))
         .Sum()
         .ToString();
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}
